lib/rules/function-component-definition.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * @fileoverview Standardize the way function component get defined
+ * @author Stefan Wullems
+ */
+
+'use strict';
+
+const Components = require('../util/Components');
+const docsUrl = require('../util/docsUrl');
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Standardize the way function component get defined',
+      category: 'Stylistic issues',
+      recommended: false,
+      url: docsUrl('function-component-definition')
+    },
+
+    schema: [{
+      enum: ['arrow-function', 'function-declaration', 'function-expression']
+    }]
+  },
+
+  create: Components.detect((context, components) => {
+    const configuration = context.options[0] || 'arrow-function';
```

Comment:
the default should not be `arrow-function`; it's a bad practice to have arrow function components. The default should probably be `function expression` - but that raises the question: should there be an option for "function declaration or expression"? It should be totally fine to have either; the primary contention is "arrow or not arrow".

Reply:
it is weird to have `function expression` as default option since it is an uncommon style.

Reply:
I think it's just about as common as function declarations; the thing that's very common is using `function`.

Reply:
@ljharb  I think there still should be an option for both function-expression as well as function declaration because at the end of the day they are different and people should still be able to prefer one over the other. Next to that I think function declarations should be the default because the react docs use those.

Reply:
I'm not objecting to there being an option for one or the other; I'm saying that the most common thing is what *does* need an option: which is "function expression or function declaration".

Reply:
Wouldn't it be better to be able to specify your preference for unnamed components? Like curried components for instance. I think it would be nice to be able to choose between arrow-expression and function-expression in that case.

Reply:
That might work for "expression", but "declaration" means it wouldn't be valid in every context.

Either way, choosing an autofixing preference is something separate; the important thing is that the rule can be configured not to warn on good things, and to warn on bad things (for whatever definition of good and bad you prefer)

Reply:
True, components with names can be arrow functions, function declarations and function expressions. Unnamed functions can only be arrow functions and function expressions. 

What I am trying to propose is that we make two things configurable that are completely separated of each other. The first one is how named components should be defined, the second one is how unnamed components should be defined. 

With named components you have the option of choosing between arrow functions, function declarations and function expressions and with unnamed components you can only choose between arrow functions and function expressions.